xstr/xstr.1
.\" Copyright (c) 1980, 1993, 1997
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)xstr.1	8.2 (Berkeley) 12/30/93
.\"
.Dd April 20, 1997
.Dt XSTR 1
.Os BSD 3
.Sh NAME
.Nm xstr
.Nd "extract strings from C programs to implement shared strings"
.Sh SYNOPSIS
.Nm xstr
.Op Fl 
.Op Fl c
.Op Fl s Ar strings
.Op Ar file
.Sh DESCRIPTION
.Nm Xstr
maintains a file
.Pa strings
into which strings in component parts of a large program are hashed.
These strings are replaced with references to this common area.
This serves to implement shared constant strings, most useful if they
are also read-only.
.Pp
Available options:
.Bl -tag -width Ds
.It Fl
.Nm Xstr
reads from the standard input.
.It Fl c
.Nm Xstr
will extract the strings from the C source
.Ar file
or the standard input
.Pq Fl ,
replacing
string references by expressions of the form (&xstr[number])
for some number.
An appropriate declaration of
.Nm xstr
is prepended to the file.
The resulting C text is placed in the file
.Pa x.c ,
to then be compiled.
The strings from this file are placed in the
.Pa strings
data base if they are not there already.
Repeated strings and strings which are suffixes of existing strings
do not cause changes to the data base.
.It Fl s Ar file
.Nm Xstr
will use
.Ar file
as
.Pa strings
data base.
.El
.Pp
After all components of a large program have been compiled a file
.Pa xs.c
declaring the common
.Nm xstr
space can be created by a command of the form
.Bd -literal -offset indent
xstr
.Ed
.Pp
The file
.Pa xs.c
should then be compiled and loaded with the rest
of the program.
If possible, the array can be made read-only (shared) saving
space and swap overhead.
.Pp
.Nm Xstr
can also be used on a single file.
A command
.Bd -literal -offset indent
xstr name
.Ed
.Pp
creates files
.Pa x.c
and
.Pa xs.c
as before, without using or affecting any
.Pa strings
file in the same directory.
.Pp
It may be useful to run
.Nm xstr
after the C preprocessor if any macro definitions yield strings
or if there is conditional code which contains strings
which may not, in fact, be needed.
An appropriate command sequence for running
.Nm xstr
after the C preprocessor is:
.Pp
.Bd -literal -offset indent -compact
cc \-E name.c | xstr \-c \-
cc \-c x.c
mv x.o name.o
.Ed
.Pp
.Nm Xstr
does not touch the file
.Pa strings
unless new items are added, thus
.Xr make 1
can avoid remaking
.Pa xs.o
unless truly necessary.
.Sh FILES
.Bl -tag -width /tmp/xsxx* -compact
.It Pa strings
Data base of strings
.It Pa x.c
Massaged C source
.It Pa xs.c
C source for definition of array `xstr'
.It Pa /tmp/xs*
Temp file when `xstr name' doesn't touch
.Pa strings
.El
.Sh SEE ALSO
.Xr mkstr 1
.Sh BUGS
If a string is a suffix of another string in the data base,
but the shorter string is seen first by
.Nm xstr
both strings will be placed in the data base, when just
placing the longer one there will do.
.Pp
Most versions of
.Nm xstr
do not deal correctly with ANSI C string concatenation.
.Sh HISTORY
The
.Nm
command, by Bill Joy, appeared in
.Bx 3.0.
.Pp
The \-s option, and a fix for ANSI C chained string constants, were added in April 1997 by David Stes.


Changes.txt

3.3.16  * Solaris change : objc uses /usr/lib/amd64 when compiling with -m64

3.3.15  * Solaris change install libraries in prefix/lib/amd64 for 64case

	  this is to avoid setting LD_LIBRARY_PATH in the shared library case
	  as /usr/lib/amd64/ or /lib/amd64 are in the default crle setup
	  (or the link /usr/lib/64 and /lib/64)

	* set the -m64 flag via EXTRA_MFLAGS

	* compile the shared libraries with -xcode=pic13 in 64bit case

3.3.14	* Fedora 29 __builtin_bswap16 fix (gcc 8.2.1)

	* move #define OBJCVERSION to config.h in src/objc (from objc1.m)

3.3.13	* FreeBSD 11.2 _Noreturn and _Alignof fix 

3.3.12  * FreeBSD 10.4 and 11.2 and DragonFly BSD sys/_types.h runetype.h
	  clang 3.4/clang 6.0/gcc 8 fixes

	* add __alignof as builtin (in addition to the 
	  already present __alignof__, __builtin_alignof and __ALIGNOF__)

	* add _Thread_local typequal for runetype.h
	  extern _Thread_local const _RuneLocale *_ThreadRuneLocale;

3.3.11	* for Solaris and RHEL, separate package "objclibs"
	  with the shared libraries (libobjcrt.so and libobjpak.so runtime)
	  those are not required currently by the compiler itself
	  applications that are built against those shared libs,
	  can have a dependency on the objclibs package (runtime) only

	* set EXTRA_MFLAGS=-fPIC in RHEL case for building shared libraries
	  (in the objclibs.spec file)

	* fix configure script to add --build-id to ldargs on RHEL
	  this is enforced by the RHEL rpmbuild for packaging the shared libs

	* add script to build IPS Solaris packages directly (without SVR4 pkg)

	* improved Solaris package that installs in /usr instead of /usr/local

	* rename OBJCDIRLIB=lib or lib64  to LIB64NAME=lib or lib64

	* fix configure to install manpages in /usr/share/man 
	  instead of /usr/man  (there is a soft 'link' on Solaris)
	  the configure script was using the old (BSD?) /usr/man path

3.3.10	* update manual.html for version 3.3.10

	* fix location objcdlso.ld dlfile case Linux 64bit (lib64 issue)

	* fix C preprocessor with -traditional-cpp broken on darwin10/gcc 4.2.1

/usr/include/secure/_stdio.h:46: error: syntax error in macro parameter list
/usr/include/secure/_stdio.h:53: error: syntax error in macro parameter list

	* add installshlib target to Makefile to install shared libraries
	  on Solaris (verified with SUN developerstudio 12.6)

	* add EXTRA_MFLAGS to Makefile's (on UNIX)
	  could be used as in: ./configure EXTRA_MFLAGS=<extra flag>

3.3.9	* fixes i686-apple-darwin17.7.0 LLVM version 9.1.0 (clang-902.0.39.1)

	* add -darwin option to objc1 to enable some keywords like macosx,ios
	  modify configure.in script to add -darwin switch
	  (reported by A.Lipinski frotz@sympatico.ca)

	* fix for secure/_stdio.h darwin17.7 has a:

	          int     (* _Nullable _close)(void *);
	  
	  define _Nullable as -gnu type qualifier
	  define _Nonnull type qualifier , in stdlib.h

                  int (* _Nonnull __compar)(const void *, const void *));

	* fix for __attribute__ __attribute__ tmpnam() in stdio.h darwin17.7

	* fix for timingsafe_bcmp in string.h
	  has __attribute__((availability(macosx,introduced=10.12.1))) 

	  posing problem of those constants 10.12.1 (release numbers)
	  current fix could be withdrawn ...
	
	* fix for system() in stdlib.h __WATCHOS_PROHIBITED __TVOS_PROHIBITED

	* use -DOBJC_BLOCKS -U__BLOCKS__ in objc driver script 

	* some documentation updates / fix DependencyGraph example
	  (example from the Brad Cox book had useless #undef modification)

3.3.8	* Solaris IPS (Image Packaging System) manifest support

	* add amd64-pc-solaris11.4 with GCC 7.3 to Platforms GCC / Solaris11.4

        * test / upgrade Solaris 11.3 to DeveloperStudio 12.6 (from 12.4)
          pkg install pkg:/developer/developerstudio-126/cc             

	* fix 'unreached statement' warning in Block.m for SUNPRO

	* add i686-apple-darwin10.8.0 to Platforms.txt

	* Apple <stdio.h> has a <secure/_stdio.h> with 
	  /usr/include/secure/_string.h:103: warning: definition of
  	  '__builtin___stpcpy_chk' 

	  add '__builtin___stpcpy_chk' support
	 -D_FORTIFY_SOURCE=0 can be used as alternative

3.3.7	* support Solaris 11.4 cc: Studio 12.6 
	  by default the compiler builds as i386-pc-solaris2.11 32 bit
	  configure for 64 bit with ./configure amd64-pc-solaris11.4 

	* add _Atomic typeword for Solaris 11.4 stddef.h
	  -D__STDC_NO_ATOMICS__ could be used as alternative

3.3.6   * some fixes for "./configure amd64-pc-solaris11.3"
	  (or ./configure x86_64-pc-solaris11.3)

	  this is the -m64 option to support building ELF64 on Solaris 11.3
	  with the Sun/SolarisStudio 12.4 compiler

          by default the compiler is built "generic" (ELF32) but can be
          used to compile code for -m64 (ELF64) or be configured as 64bit by
	  ./configure x86_64-pc-solaris11.3

3.3.5   * add objc.SlackBuild script for building a Slackware package

3.3.4   * compiler driver fixes for SUN studio C
	Motif compile (imake) on Solaris11 support -z, -x, -B, -M
	support options -xO4 -xbuiltin=%all -xlibmil -xstrconst -xpentium -xregs=no%frameptr -Xa  -z lazyload -z combreloc -xstrconst -xildoff -B direct -M/usr/lib/ld/map (pass through those options, so that imake compile options work)

	* gcc 5 fix use 'gcc -E -x c' as cpp driver instead of "gcc -E"
	  ie. use gcc in 'C' mode rather than their 'Objective-C' mode
	  alternative would be to rename .m to .c (as CPPWANTSDOTC=Y on IRIX)
	  both work, observed on gcc 5 issue with -O2 which caused
	  the -D__GNUC_STDC_INLINE__ issue for the bits/string.h header
	  (the attribute(gnu_inline) not being emitted in Objective-C
	  ie. difference in output when invoking gcc -O2 -E or gcc -O2 -E -x c
	  they generate different code

3.3.3	* compile on Slackware 14.2 with gcc 5.3 and clang 3.8
	  NOTE: set OBJCOPT to -D__GNUC_STDC_INLINE__ for bootstrap
          cfr. gcc 5.2 with dragonfly 4.4.1  -D__GNUC_STDC_INLINE__

	* fix builtin type __builtin_offsetof for Motif X compile of richtext


xstr/xstr.m
/*
 * Copyright (c) 1980, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static char copyright[] =
"@(#) Copyright (c) 1980, 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)xstr.c	8.1 (Berkeley) 6/9/93";
#endif /* not lint */

#include <sys/types.h>
#include <signal.h>
#include <unistd.h> /* mktemp() */
#include <errno.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include "pathnames.h"  /* seems overkill for just one define... */

#include "objpak.h"

/*
 * xstr - extract and hash strings in a C program
 *
 * Bill Joy, UCB
 * November, 1978
 */

/*
 * David Stes,
 * April, 1997
 * 
 * (a) Added -s option so that strings databases can be shared
 *     across directories.
 * 
 * (b) Fix for ANSI C string concatenation.
 *     "hello" "world" was being translated into 2 separate &xstr expressions.
 */

#define	ignore(a)	((void) a)

off_t	tellpt;
off_t	hashit();
void	onintr();
char	*savestr();
id		yankstr(); /* stes 4/21/1997 */
off_t	yankchain(); /* stes 4/21/1997 - replaces yankstr() */

off_t	mesgpt;
char	*strings =	"strings";

int	cflg;
int	vflg;
int	readstd;
int tmpflg;   /* flag temporary "strings" file for cleanup */

char *umsg = "usage: xstr [ -v ] [ -c ] [ -s strings ] [ - ] [ name ... ]\n";

usage()
{
	fprintf(stderr,umsg);exit(1);
}

void
mktempstrings()
{
	static char template[256];
	strcpy(template,_PATH_TMP);
	strings = mktemp(template);
	tmpflg=1;
}

isoption(s)
	char *s;
{
	return s != NULL && s[0] == '-';
}

main(argc, argv)
	int argc;
	char *argv[];
{

	argc--, argv++;
	while (argc > 0 && isoption(*argv)) {
		register char *cp = &(*argv++)[1];

		argc--;
		if (*cp == 0) {
			readstd++;
			continue;
		}
		do switch (*cp++) {

		case 'c':
			cflg++;
			continue;

		case 'v':
			vflg++;
			continue;
		
		/* stes 4/21/1997 -- use strings databases across directories */
		case 's':
			argc--;strings=*argv++;
			if (!strings || isoption(strings)) 	usage();
			continue;

		default:
			usage();

		} while (*cp);
	}
	
	if (signal(SIGINT, SIG_IGN) == SIG_DFL) signal(SIGINT, onintr);
	
	if (cflg || (argc == 0 && !readstd))
		readstrings();
	else
		mktempstrings();
	
	while (readstd || argc > 0) {
		if (freopen("x.c", "w", stdout) == NULL)
			perror("x.c"), exit(1);
		if (!readstd && freopen(argv[0], "r", stdin) == NULL)
			perror(argv[0]), exit(2);
		process("x.c");
		if (readstd == 0)
			argc--, argv++;
		else
			readstd = 0;
	};
	flushsh();
	if (cflg == 0)
		xsdotc();
	if (tmpflg) ignore(unlink(strings));
	exit(0);
}

char linebuf[BUFSIZ];

int
newlinebuf(char *msg)
{
	if (fgets(linebuf, sizeof linebuf, stdin) == NULL) {
		if (ferror(stdin)) { perror(msg);exit(3); } else return 1;
	}
	return 0;
}

/*
 * stes - 21/6/1997
 * Return String instance for preprocessor line.
 * Code copied out of process() and reused in whitespace scanning
 */

id
cppline()
{
	if (linebuf[0] == '#') {
		if (linebuf[1] == ' ' && isdigit(linebuf[2]))
			return [String sprintf:"#line%s", &linebuf[1]];
		else
			return [String sprintf:"%s", linebuf];
	}

	return nil;
}

process(name)
	char *name;
{
	char *cp;
	register int c;
	register int incomm = 0;
	int inchain = 0;
	int ret;

	printf("extern char\txstr[];\n");
	for (;;) {
		id str;

		if (newlinebuf(name)) break;
		if (str = cppline()) { [str print];[str free];continue; }
		
		for (cp = linebuf; c = *cp++;) switch (c) {

		case '"':
			if (incomm)
				goto def;
			if ((ret = (int) yankchain(&cp)) == -1)
				goto out;
			printf("(&xstr[%d])", ret);
			break;

		case '\'':
			if (incomm)
				goto def;
			putchar(c);
			if (*cp)
				putchar(*cp++);
			break;

		case '/':
			if (incomm || *cp != '*')
				goto def;
			incomm = 1;
			cp++;
			printf("/*");
			continue;

		case '*':
			if (incomm && *cp == '/') {
				incomm = 0;
				cp++;
				printf("*/");
				continue;
			}
			goto def;

def:
		default:
			putchar(c);
			break;
		}
	}
out:
	if (ferror(stdout))
		perror("x.c"), onintr();
}

/* stes - 4/21/1997
 * modified to return a String instance instead of an off_t.
 * yankchain() then concatenates string objects.
 */


id
yankstr(cpp)
	register char **cpp;
{
	register char *cp = *cpp;
	register int c, ch;
	char dbuf[BUFSIZ];
	register char *dp = dbuf;
	register char *tp;

	while (c = *cp++) {
		switch (c) {

		case '"':
			cp++;
			goto out;

		case '\\':
			c = *cp++;
			if (c == 0)
				break;
			if (c == '\n') {
				if (newlinebuf("x.c")) return nil;
				cp = linebuf;
				continue;
			}
			for (tp = "b\bt\tr\rn\nf\f\\\\\"\""; ch = *tp++; tp++)
				if (c == ch) {
					c = *tp;
					goto gotc;
				}
			if (!octdigit(c)) {
				*dp++ = '\\';
				break;
			}
			c -= '0';
			if (!octdigit(*cp))
				break;
			c <<= 3, c += *cp++ - '0';
			if (!octdigit(*cp))
				break;
			c <<= 3, c += *cp++ - '0';
			break;
		}
gotc:
		*dp++ = c;
	}
out:
	*cpp = --cp;
	*dp = 0;
	return [String str:dbuf]; /* return (hashit(dbuf, 1)); */
}

/* stes - 4/21/1997
 * skip white space, until a double quote follows
 * skip white space if yes, otherwise output whitespace.
 * (this is complicated by the use of linebuf)
 */

void
concatChar(id str,int c)
{
	char cStr[2];
	cStr[0] = c;cStr[1] = 0;
	[str concatSTR:cStr];
}

int
stringfollows(cpp)
	char **cpp;
{
	int c;
	id str; /* for stripping #line directives out of string */
	char *cp = *cpp;
	id whitespace = [String new];

	while ((c = *cp++)) switch(c) {
		case '\n' :
				concatChar(whitespace,c);
				do {
				if (newlinebuf("x.c")) break;else cp = linebuf;
				if ((str = cppline())) {
					[whitespace concatSTR:[str str]];
					[str free];
				}} while (str);
				continue;
		case ' '  :
		case '\t' :
				concatChar(whitespace,c);continue;
		case '"' :
				*cpp = cp;return 1;
		default   :
				*cpp = --cp;[whitespace print];return 0;
	}
}

/* stes - 4/21/1997
 * yankchain() is like the old yankstr() but deals with
 * ANSI C string concatenation.
 */

off_t
yankchain(cpp)
	char **cpp;
{
	id chain = [String new];
	
	do {
		id str = yankstr(cpp);
		if (str) [chain concatSTR:[str str]]; else return -1;
	} while (stringfollows(cpp));
	
	return (hashit([chain str], 1));
}

octdigit(c)
	char c;
{

	return (isdigit(c) && c != '8' && c != '9');
}

fgetNUL(obuf, rmdr, file)
	char *obuf;
	register int rmdr;
	FILE *file;
{
	register c;
	register char *buf = obuf;

	while (--rmdr > 0 && (c = xgetc(file)) != 0 && c != EOF)
		*buf++ = c;
	*buf++ = 0;
	return ((feof(file) || ferror(file)) ? 0 : 1);
}

readstrings()
{
	char buf[BUFSIZ];
	register FILE *mesgread = fopen(strings, "r");

	if (mesgread == NULL)
		return;
	for (;;) {
		mesgpt = tellpt;
		if (fgetNUL(buf, sizeof buf, mesgread) == NULL) break;
		ignore(hashit(buf, 0));
	}
	ignore(fclose(mesgread));
}

xgetc(file)
	FILE *file;
{

	tellpt++;
	return (getc(file));
}

#define	BUCKETS	128

struct	hash {
	off_t	hpt;
	char	*hstr;
	struct	hash *hnext;
	short	hnew;
} bucket[BUCKETS];

off_t
hashit(str, new)
	char *str;
	int new;
{
	int i;
	register struct hash *hp, *hp0;

	hp = hp0 = &bucket[lastchr(str) & 0177];
	while (hp->hnext) {
		hp = hp->hnext;
		i = istail(str, hp->hstr);
		if (i >= 0)
			return (hp->hpt + i);
	}
	if ((hp = (struct hash *) calloc(1, sizeof (*hp))) == NULL) {
		perror("xstr");
		exit(8);
	}
	hp->hpt = mesgpt;
	if (!(hp->hstr = strdup(str))) {
		(void)fprintf(stderr, "xstr: %s\n", strerror(errno));
		exit(1);
	}
	mesgpt += strlen(hp->hstr) + 1;
	hp->hnext = hp0->hnext;
	hp->hnew = new;
	hp0->hnext = hp;
	return (hp->hpt);
}

flushsh()
{
	register int i;
	register struct hash *hp;
	register FILE *mesgwrit;
	register int old = 0, new = 0;

	for (i = 0; i < BUCKETS; i++)
		for (hp = bucket[i].hnext; hp != NULL; hp = hp->hnext)
			if (hp->hnew)
				new++;
			else
				old++;
	if (new == 0 && old != 0)
		return;
	mesgwrit = fopen(strings, old ? "r+" : "w");
	if (mesgwrit == NULL)
		perror(strings), exit(4);
	for (i = 0; i < BUCKETS; i++)
		for (hp = bucket[i].hnext; hp != NULL; hp = hp->hnext) {
			found(hp->hnew, hp->hpt, hp->hstr);
			if (hp->hnew) {
				fseek(mesgwrit, hp->hpt, 0);
				ignore(fwrite(hp->hstr, strlen(hp->hstr) + 1, 1, mesgwrit));
				if (ferror(mesgwrit))
					perror(strings), exit(4);
			}
		}
	if (fclose(mesgwrit) == EOF)
		perror(strings), exit(4);
}

found(new, off, str)
	int new;
	off_t off;
	char *str;
{
	if (vflg == 0)
		return;
	if (!new)
		fprintf(stderr, "found at %d:", (int) off);
	else
		fprintf(stderr, "new at %d:", (int) off);
	prstr(str);
	fprintf(stderr, "\n");
}

prstr(cp)
	register char *cp;
{
	register int c;

	while (c = (*cp++ & 0377))
		if (c < ' ')
			fprintf(stderr, "^%c", c + '`');
		else if (c == 0177)
			fprintf(stderr, "^?");
		else if (c > 0200)
			fprintf(stderr, "\\%03o", c);
		else
			fprintf(stderr, "%c", c);
}

xsdotc()
{
	register FILE *strf = fopen(strings, "r");
	register FILE *xdotcf;

	if (strf == NULL)
		perror(strings), exit(5);
	xdotcf = fopen("xs.c", "w");
	if (xdotcf == NULL)
		perror("xs.c"), exit(6);
	fprintf(xdotcf, "char\txstr[] = {\n");
	for (;;) {
		register int i, c;

		for (i = 0; i < 8; i++) {
			c = getc(strf);
			if (ferror(strf)) {
				perror(strings);
				onintr();
			}
			if (feof(strf)) {
				fprintf(xdotcf, "\n");
				goto out;
			}
			fprintf(xdotcf, "0x%02x,", c);
		}
		fprintf(xdotcf, "\n");
	}
out:
	fprintf(xdotcf, "};\n");
	ignore(fclose(xdotcf));
	ignore(fclose(strf));
}

lastchr(cp)
	register char *cp;
{

	while (cp[0] && cp[1])
		cp++;
	return (*cp);
}

istail(str, of)
	register char *str, *of;
{
	register int d = strlen(of) - strlen(str);

	if (d < 0 || strcmp(&of[d], str) != 0)
		return (-1);
	return (d);
}

void
onintr()
{

	ignore(signal(SIGINT, SIG_IGN));
	if (tmpflg) ignore(unlink(strings));
	ignore(unlink("x.c"));
	ignore(unlink("xs.c"));
	exit(7);
}
